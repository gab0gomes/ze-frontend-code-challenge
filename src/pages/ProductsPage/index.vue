<template>
  <main class="container">
    <spinner v-if="toggleSpinner" />
    <state v-else-if="!params" :src="mapImage" alt-text="uma figura de um mapa">
      <p>Nenhum endereço selecionado.</p>
      <p>
        <router-link to="/"
          >Clique aqui para selecionar um endereço.</router-link
        >
      </p>
    </state>
    <state
      v-else-if="toggleNoPocState"
      :src="shopImage"
      alt-text="uma figura de um supermercado"
    >
      <p>Desculpe, mas não atendemos nessa área ainda... :(</p>
      <p>
        Talvez você queira
        <router-link to="/">escolher outro endereço?</router-link>
      </p>
    </state>
    <state
      v-else-if="poc && !poc.products.length"
      :src="boxImage"
      alt-text="uma figura de uma caixa de papelão vazia"
    >
      <p>Nenhum produto encontrado! :(</p>
      <p>
        Talvez você queira
        <router-link to="/">escolher outro endereço?</router-link>
      </p>
    </state>
    <div v-else-if="categorizedProducts">
      <product-grid
        v-for="(products, category) in categorizedProducts"
        :key="category"
        :title="category"
        :products="products"
      />
    </div>
  </main>
</template>

<script>
import gql from 'graphql-tag'
import { mapState } from 'vuex'

import pocSearchQuery from '@/core/graphQL/queries/pocSearch'
import pocQuery from '@/core/graphQL/queries/poc'

import State from './components/StateComponent'
import ProductGrid from './components/ProductsGridComponent'
import Spinner from '@/core/components/SpinnerComponent'

import mapImage from '../../../static/images/map.png'
import shopImage from '../../../static/images/shop.png'
import boxImage from '../../../static/images/box.png'

export default {
  components: {
    State,
    ProductGrid,
    Spinner
  },

  data() {
    return {
      mapImage,
      shopImage,
      boxImage,
      pocSearch: '',
      poc: ''
    }
  },

  apollo: {
    pocSearch: {
      query: gql`
        ${pocSearchQuery}
      `,
      variables() {
        return this.params
      },
      skip() {
        return !this.params
      }
    },
    poc: {
      query: gql`
        ${pocQuery}
      `,
      variables() {
        return {
          id: this.pocSearch[0]?.id,
          search: '',
          categoryId: null
        }
      },
      skip() {
        return !this.pocSearch[0]?.id
      }
    }
  },

  computed: {
    ...mapState('address', {
      location: (state) => state.location
    }),

    params() {
      if (this.location.lat && this.location.lng) {
        return {
          algorithm: 'NEAREST',
          lat: `${this.location.lat}`,
          long: `${this.location.lng}`,
          now: new Date().toISOString()
        }
      }

      return undefined
    },

    toggleNoPocState() {
      return typeof this.pocSearch !== 'string' && !this.pocSearch.length
    },

    toggleSpinner() {
      return this.$apollo.loading
    },

    categorizedProducts() {
      if (this.poc.products) {
        const groups = this.poc.products.reduce((acc, product) => {
          let key = product.productVariants[0].subtitle || 'Outros'
          const accumulator = { ...acc }

          if (accumulator[key]) {
            accumulator[key].push(product)
            accumulator[key] = this.sortArrayByKey(accumulator[key], 'title')
          } else {
            accumulator[key] = [product]
          }

          return accumulator
        }, {})

        return this.sortObjectByKeys(groups)
      }
      return {}
    }
  },

  methods: {
    sortObjectByKeys(unsorted) {
      return Object.keys(unsorted)
        .sort()
        .reduce((obj, key) => {
          obj[key] = unsorted[key]
          return obj
        }, {})
    },

    sortArrayByKey(unsortedArray, key) {
      return unsortedArray.sort((a, b) => {
        if (a[key] > b[key]) {
          return 1
        }
        if (a[key] < b[key]) {
          return -1
        }
        return 0
      })
    }
  }
}
</script>

<style scoped>
main {
  min-height: 100%;
}

.container {
  display: flex;
  justify-content: center;
  align-items: center;
}
</style>
