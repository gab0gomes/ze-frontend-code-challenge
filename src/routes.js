import HomePage from '@/pages/Home'

const routes = [
  {
    path: '/',
    name: 'Home',
    component: HomePage
  }
]

export default routes
