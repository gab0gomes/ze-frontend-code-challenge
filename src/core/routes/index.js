import HomePage from '@/pages/HomePage'
import ProductsPage from '@/pages/ProductsPage'

const routes = [
  {
    path: '/',
    name: 'Home',
    component: HomePage
  },
  {
    path: '/products',
    name: 'Products',
    component: ProductsPage
  }
]

export default routes
