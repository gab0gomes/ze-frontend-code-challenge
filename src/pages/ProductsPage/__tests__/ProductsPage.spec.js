import { mount, createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import flushPromises from 'flush-promises'
import VueRouter from 'vue-router'

import ProductsPage from '../index'

import addressStore from '@/core/store/address'

import routes from '@/core/routes'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)

const store = (withAddress = true) =>
  new Vuex.Store({
    modules: {
      address: {
        ...addressStore,
        ...(withAddress
          ? {
              state: {
                fancyAddress: 'R. Fradique Coutinho, 1632 - Pinheiros',
                location: {
                  lat: -23.5536452,
                  lng: -46.694965
                }
              }
            }
          : {})
      }
    }
  })

const router = () =>
  new VueRouter({
    routes,
    linkActiveClass: 'active',
    mode: 'abstract'
  })

jest.mock('@/core/services/geocoding')

const data = (withPoc = true, withProducts = true) => ({
  ...(withProducts
    ? {
        poc: {
          id: '532',
          products: [
            {
              id: '8868',
              title: 'Skol 269ml - Unidade',
              rgb: false,
              images: [
                {
                  url:
                    'https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/product/8868_205f958d-2e51-48a3-b4d5-a2998765571a.jpg',
                  __typename: 'ProductImage'
                }
              ],
              productVariants: [
                {
                  availableDate: '2018-10-31T00:00:00',
                  productVariantId: '8502',
                  price: 2.09,
                  inventoryItemId: '80149',
                  shortDescription: null,
                  title: 'Skol 269ml - Unidade',
                  published: null,
                  volume: '0,00269',
                  volumeUnit: null,
                  description: null,
                  subtitle: 'Cervejas',
                  components: [],
                  __typename: 'ProductVariant'
                }
              ],
              __typename: 'Product'
            },
            {
              id: '8869',
              title: 'Skol 269ml - Pack com 15 Unidades',
              rgb: false,
              images: [
                {
                  url:
                    'https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/products/8503.png',
                  __typename: 'ProductImage'
                }
              ],
              productVariants: [
                {
                  availableDate: '2018-10-31T00:00:00',
                  productVariantId: '8503',
                  price: 31.35,
                  inventoryItemId: '80151',
                  shortDescription: null,
                  title: 'Skol 269ml - Pack com 15 Unidades',
                  published: null,
                  volume: '0,04035',
                  volumeUnit: null,
                  description: null,
                  subtitle: 'Cervejas',
                  components: [],
                  __typename: 'ProductVariant'
                }
              ],
              __typename: 'Product'
            },
            {
              id: '8877',
              title: 'Brahma 269ml - Unidade',
              rgb: false,
              images: [
                {
                  url:
                    'https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/product/00008877_02c7ebe9-c23a-405c-8af8-8f0988f276c7.jpg',
                  __typename: 'ProductImage'
                }
              ],
              productVariants: [
                {
                  availableDate: '2018-10-31T00:00:00',
                  productVariantId: '8511',
                  price: 1.99,
                  inventoryItemId: '80139',
                  shortDescription: null,
                  title: 'Brahma 269ml - Unidade',
                  published: null,
                  volume: '0,00269',
                  volumeUnit: null,
                  description: null,
                  subtitle: null,
                  components: [],
                  __typename: 'ProductVariant'
                }
              ],
              __typename: 'Product'
            },
            {
              id: '9214',
              title: 'Guaraná 350ml - Pack com 12 Unidades',
              rgb: false,
              images: [
                {
                  url:
                    'https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/product/00009214_9e6b60c7-b7f2-47d0-aecd-bb702a9a07b0.jpg',
                  __typename: 'ProductImage'
                }
              ],
              productVariants: [
                {
                  availableDate: '2018-10-31T00:00:00',
                  productVariantId: '8848',
                  price: 29.88,
                  inventoryItemId: '80185',
                  shortDescription: null,
                  title: 'Guaraná 350ml - Pack com 12 Unidades',
                  published: null,
                  volume: '0,042',
                  volumeUnit: null,
                  description: null,
                  subtitle: 'Não Alcóolicos',
                  components: [],
                  __typename: 'ProductVariant'
                }
              ],
              __typename: 'Product'
            }
          ],
          __typename: 'POC'
        }
      }
    : { poc: { id: '532', __typename: 'POC', products: [] } }),
  ...(withPoc
    ? { pocSearch: [{ id: '532', __typename: 'POC' }] }
    : { pocSearch: [] })
})

describe('PoductsPage', () => {
  it('render correctly', async () => {
    const wrapper = mount(ProductsPage, {
      localVue,
      store: store(),
      router: router(),
      mocks: {
        $apollo: {
          loading: false
        }
      }
    })

    wrapper.setData({
      ...data()
    })

    await flushPromises()

    expect(wrapper.element).toMatchSnapshot()

    wrapper.destroy()
  })

  it('should show a spinner if graphql is loading', async () => {
    const wrapper = mount(ProductsPage, {
      localVue,
      store: store(),
      router: router(),
      mocks: {
        $apollo: {
          loading: true
        }
      }
    })

    wrapper.setData({
      ...data()
    })

    await flushPromises()

    expect(wrapper.findAll('.spinner-container').length).toBe(1)

    wrapper.destroy()
  })

  it('should show a no address state if none address was provided', async () => {
    const wrapper = mount(ProductsPage, {
      localVue,
      store: store(false),
      router: router(),
      mocks: {
        $apollo: {
          loading: false
        }
      }
    })

    wrapper.setData({
      ...data()
    })

    await flushPromises()

    expect(wrapper.text().includes('Nenhum endereço selecionado.')).toBe(true)

    wrapper.destroy()
  })

  it('should show a no poc state if none poc is returned', async () => {
    const wrapper = mount(ProductsPage, {
      localVue,
      store: store(),
      router: router(),
      mocks: {
        $apollo: {
          loading: false
        }
      }
    })

    wrapper.setData({
      ...data(false)
    })

    await flushPromises()

    expect(
      wrapper
        .text()
        .includes('Desculpe, mas não atendemos nessa área ainda... :(')
    ).toBe(true)

    wrapper.destroy()
  })

  it('should show a no products state if poc.products is empty', async () => {
    const wrapper = mount(ProductsPage, {
      localVue,
      store: store(),
      router: router(),
      mocks: {
        $apollo: {
          loading: false
        }
      }
    })

    wrapper.setData({
      ...data(true, false)
    })

    await flushPromises()

    expect(wrapper.text().includes('Nenhum produto encontrado! :(')).toBe(true)

    wrapper.destroy()
  })

  it('should categorize and sort products', async () => {
    const wrapper = mount(ProductsPage, {
      localVue,
      store: store(),
      router: router(),
      mocks: {
        $apollo: {
          loading: false
        }
      }
    })

    wrapper.setData({
      ...data()
    })

    await flushPromises()

    const expectedData = {
      Cervejas: [
        { title: 'Skol 269ml - Pack com 15 Unidades' },
        { title: 'Skol 269ml - Unidade' }
      ],
      'Não Alcóolicos': [{ title: 'Guaraná 350ml - Pack com 12 Unidades' }],
      Outros: [{ title: 'Brahma 269ml - Unidade' }]
    }

    expect(wrapper.vm.categorizedProducts).toMatchObject(expectedData)

    wrapper.destroy()
  })
})
