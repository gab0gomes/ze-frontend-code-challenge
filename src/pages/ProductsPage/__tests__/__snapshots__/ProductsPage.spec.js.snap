// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoductsPage render correctly 1`] = `
<main
  class="container"
>
  <div>
    <div
      class="cart-container"
    >
      <img
        alt="uma figura de um carrinho de compras"
        src="static/images/cart.png"
      />
       
      <p>
        
    R$ 0,00
  
      </p>
    </div>
     
    <section
      class="products-grid"
    >
      <h2>
        Cervejas
      </h2>
       
      <hr />
       
      <div
        class="grid-container"
      >
        <div
          class="product-card"
        >
          <div
            class="product-description"
          >
            <!---->
             
            <img
              src="https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/products/8503.png"
            />
             
            <h5>
              Skol 269ml - Pack com 15 Unidades
            </h5>
          </div>
           
          <div
            class="product-shop"
          >
            <p
              class="price"
            >
              R$ 31,35
            </p>
             
            <div
              class="controls"
            >
              <button
                alt="Remover uma unidade do carrinho"
                class="button remove"
              >
                
        -
      
              </button>
               
              <button
                alt="Adicionar uma unidade ao carrinho"
                class="button add"
              >
                
        +
      
              </button>
            </div>
          </div>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-description"
          >
            <!---->
             
            <img
              src="https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/product/8868_205f958d-2e51-48a3-b4d5-a2998765571a.jpg"
            />
             
            <h5>
              Skol 269ml - Unidade
            </h5>
          </div>
           
          <div
            class="product-shop"
          >
            <p
              class="price"
            >
              R$ 2,09
            </p>
             
            <div
              class="controls"
            >
              <button
                alt="Remover uma unidade do carrinho"
                class="button remove"
              >
                
        -
      
              </button>
               
              <button
                alt="Adicionar uma unidade ao carrinho"
                class="button add"
              >
                
        +
      
              </button>
            </div>
          </div>
        </div>
      </div>
    </section>
    <section
      class="products-grid"
    >
      <h2>
        Não Alcóolicos
      </h2>
       
      <hr />
       
      <div
        class="grid-container"
      >
        <div
          class="product-card"
        >
          <div
            class="product-description"
          >
            <!---->
             
            <img
              src="https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/product/00009214_9e6b60c7-b7f2-47d0-aecd-bb702a9a07b0.jpg"
            />
             
            <h5>
              Guaraná 350ml - Pack com 12 Unidades
            </h5>
          </div>
           
          <div
            class="product-shop"
          >
            <p
              class="price"
            >
              R$ 29,88
            </p>
             
            <div
              class="controls"
            >
              <button
                alt="Remover uma unidade do carrinho"
                class="button remove"
              >
                
        -
      
              </button>
               
              <button
                alt="Adicionar uma unidade ao carrinho"
                class="button add"
              >
                
        +
      
              </button>
            </div>
          </div>
        </div>
      </div>
    </section>
    <section
      class="products-grid"
    >
      <h2>
        Outros
      </h2>
       
      <hr />
       
      <div
        class="grid-container"
      >
        <div
          class="product-card"
        >
          <div
            class="product-description"
          >
            <!---->
             
            <img
              src="https://courier-images-codechallenge.s3-us-west-2.amazonaws.com/product/00008877_02c7ebe9-c23a-405c-8af8-8f0988f276c7.jpg"
            />
             
            <h5>
              Brahma 269ml - Unidade
            </h5>
          </div>
           
          <div
            class="product-shop"
          >
            <p
              class="price"
            >
              R$ 1,99
            </p>
             
            <div
              class="controls"
            >
              <button
                alt="Remover uma unidade do carrinho"
                class="button remove"
              >
                
        -
      
              </button>
               
              <button
                alt="Adicionar uma unidade ao carrinho"
                class="button add"
              >
                
        +
      
              </button>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</main>
`;
